node/src/keyshare/permanent.rs

```diff
@@ -0,0 +1,222 @@
+use super::Keyshare;
+use anyhow::Context;
+use k256::{AffinePoint, Scalar};
+use mpc_contract::primitives::key_state::EpochId;
+use serde::{Deserialize, Serialize};
+
+/// The single object we persist to permanent key storage.
+/// Corresponds to a Keyset in the contract side (i.e. one keyshare per domain).
+#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
+pub struct PermanentKeyshareData {
+    pub epoch_id: EpochId,
+    /// These keyshares are in the exact same order as the domains in the Keyset.
+    pub keyshares: Vec<Keyshare>,
+}
+
+impl PermanentKeyshareData {
+    pub fn from_legacy(legacy: &LegacyRootKeyshareData) -> Self {
+        Self {
+            epoch_id: EpochId::new(legacy.epoch),
+            keyshares: vec![Keyshare::from_legacy(legacy)],
+        }
+    }
+}
+
+/// Backend for storing the permanent keyshare data.
+#[async_trait::async_trait]
+pub trait PermanentKeyStorageBackend: Send + Sync {
+    /// Loads the latest stored data; or None if no data was ever stored.
+    async fn load(&self) -> anyhow::Result<Option<Vec<u8>>>;
+    /// Stores the data, making it the latest. The identifier is used for the local backend to write
+    /// the data to a separate file and then linking it to the main file (acting as a backup
+    /// mechanism).
+    async fn store(&self, data: &[u8], identifier: &str) -> anyhow::Result<()>;
```

Comment:
We fail in local if the given identifier already exists, whereas in gcp we do not.
Because we check the existence of the file on the `PermamentKeyStore` level, it should not be an issue in practice, but maybe it would be good specify this in the comment.
Or maybe change the behavior of the local storage as well and rename this function `store_or_override` or sth similar.

Reply:
Good catch! I wanted to tag it with a name in GCP too but GCP doesn't let me tag a secret version with anything at all. I don't really want to override it though, coz once it has the ability to override anything, there's the possibility of corrupting something useful. So I think the issue is really that on the filesystem we don't have a "append" mechanism, whereas on GCP we do. But since the collision on disk is expected to never happen (unless there's a bug), I think that code can be considered unreachable and thus the discrepancy in the behavior can be ignored.